File: go/euler23.go
package main
import "fmt"

func properDivisorSum(n int) int {
    if n == 1 {
        return 0
    }
    res := 1
    for i := 2; i * i <= n; i++ {
        if n % i == 0 {
            res += i
            if n / i != i {
                res += n / i
            }
        }
    }
    return res
}

func main() {
    res := 0
    limit := 28123
    abNums := []int{}
    for i := 1; i <= limit; i++ {
        if properDivisorSum(i) > i {
            abNums = append(abNums, i)
        }
    }
    
    abSums := make([]bool, limit + 1)
    for i := 0; i < len(abNums); i++ {
        for j := i; j < len(abNums); j++ {
            sum := abNums[i] + abNums[j]
            if sum > limit {
                break
            }
            abSums[sum] = true
        }
    }
    for i := 1; i <= limit; i++ {
        if !abSums[i] {
            res += i
        }
    }
    fmt.Println(res)
}

File: ts/euler23.ts
function properDivisorSum(n: number): number {
    let res: number = 1;
    for (let i: number = 2; i * i <= n; i++) {
        if (n % i == 0) {
            res += i;
            if (n / i != i) {
                res += n / i;
            }
        }
    }
    return res;
}

function nonAbundantSums(): number {
    let n: number = 28123;
    // precompute sum of all numbers up to 28123
    let res: number = n * (n + 1) / 2;
    let abNums: Set<number> = new Set<number>();
    for (let i: number = 12; i <= 28123; i++) {
        // check if i is an abundant number
        if (properDivisorSum(i) > i) abNums.add(i);
        // check if i is a sum of two abundant numbers
        for (let abNum of abNums) {
            // i is an abundant sum, subtract from total
            if (abNums.has(i - abNum)) {
                res -= i;
                break;
            }
        }
    }
    return res;
}

console.log(nonAbundantSums());
